### ISSUE TITLE
Streaming reading in wasmparser

### ISSUE THREAD

**Issue by evmar** [bytecodealliance/wasm-tools#328]
I looked at wasmparser and naively expected the API to work with a `std::io::BufReader` as input, rather than requiring the full file contents up front.

Is there something fundamental to the WASM format that precludes this?  Was it an intentional design decision here to not use such a thing?  Glancing at `binary_reader`, it seems it works with a cursor-like concept anyway.

**Comment by alexcrichton** [comment]
There's an [example of streaming reading in the documentation](https://docs.rs/wasmparser/0.80.1/wasmparser/struct.Parser.html#examples), does that help?

Direct support isn't added in this crate right now because somewhat opinionated decisions about buffering need to be made, and for now it's figured that's best left up to applications.

**Comment by evmar** [comment]
Yes, thank you!  Sorry for not finding that documentation section.

Retracting my steps, I think from [here](https://github.com/bytecodealliance/wasm-tools/tree/main/crates/wasmparser#documentation) the link drops you into [here](https://docs.rs/wasmparser/0.80.1/wasmparser/) but from that you have to scan down and guess that `Parser` is the entry point.  Maybe call that out in the doc overview?

**Comment by alexcrichton** [comment]
No problem! The documentation of `wasmparser` is definitely one that could use a little love. I'll leave this open to track improving the `wasmparser` docs.
